packages/backend/src/peripherals/starkware/L2TransactionDownloader.ts

```diff
@@ -0,0 +1,56 @@
+import { Logger } from '../../tools/Logger'
+import { L2TransactionRepository } from '../database/L2TransactionRepository'
+import { StateUpdateRecord } from '../database/StateUpdateRepository'
+import { FeederGatewayClient } from './FeederGatewayClient'
+
+export class L2TransactionDownloader {
+  constructor(
+    private readonly feederGatewayClient: FeederGatewayClient,
+    private readonly transactionRepository: L2TransactionRepository,
+    private readonly logger: Logger
+  ) {}
+
+  async sync(
+    stateUpdates: Pick<StateUpdateRecord, 'id' | 'batchId' | 'blockNumber'>[]
+  ) {
+    const latestSyncedTransactionStateUpdateId =
+      (await this.transactionRepository.findLatestStateUpdateId()) ?? 0
+
+    for (const stateUpdate of stateUpdates) {
+      if (stateUpdate.id <= latestSyncedTransactionStateUpdateId) {
+        continue
+      }
+      this.logger.info(`Syncing transactions...`, {
+        stateUpdateId: stateUpdate.id,
+      })
+
+      const data = await this.feederGatewayClient.getPerpetualBatchInfo(
```

Comment:
I'm pretty sure sometimes feederGatway will not have the data for the stateUpdateId we ask for. We need to handle this case gracefully and stop the collection process.